include/seqan3/core/algorithm/concept.hpp

```diff
@@ -0,0 +1,150 @@
+#include <seqan3/std/concept/callable.hpp>
+#include <seqan3/std/concept/core_language.hpp>
+#include <seqan3/std/concept/object.hpp>
+
+namespace seqan3::detail
+{
+
+// ----------------------------------------------------------------------------
+// Concept config_element_concept
+// ----------------------------------------------------------------------------
+
+//!\cond
+template <typename derived_t>
+class config_element_base;
+//!\endcond
+
+/*!\interface seqan3::detail::config_element_concept <>
+ * \brief Concept for an algorithm configuration.
+ * \ingroup core_algorithm
+ *
+ * \extends seqan3::default_constructible_concept
+ * \extends seqan3::assignable_concept
+ * \implements seqan3::detail::config_element_base
+ *
+ * \details
+ *
+ * Classes must extend the abstract crtp-base class seqan3::detail::config_element_base in order to satisfy this
+ * concept. This concept is merely used for internal purposes and shall not be exposed in public user interfaces.
+ */
+//!\cond
+template <typename config_t>
+concept bool config_element_concept = default_constructible_concept<std::remove_reference_t<config_t>> &&
+                                      assignable_concept<std::remove_reference_t<config_t> &, config_t> &&
+                                      std::is_base_of_v<config_element_base<std::remove_reference_t<config_t>>,
+                                                        std::remove_reference_t<config_t>>;
+//!\endcond
+
+// ----------------------------------------------------------------------------
+// Concept deferred_config_element_concept
+// ----------------------------------------------------------------------------
+
+//!\cond
+// Forward declaration for the deferred config concept
+template <detail::config_element_concept ... configs_t>
+class configuration;
+
+template <typename derived_t>
+class deferred_config_element_base;
+
+auto _callable_dummy = [] (auto && cfg)
+{
+    return std::forward<decltype(cfg)>(cfg);
+};
+//!\endcond
+
+/*!\interface seqan3::detail::deferred_config_element_concept <>
+ * \brief Concept for a deferred algorithm configuration.
+ * \ingroup core_algorithm
+ *
+ * \extends seqan3::invocable_concept
+ * \implements seqan3::detail::deferred_config_element_base
+ *
+ * \details
+ *
+ * Classes must extend the abstract crtp-base class seqan3::detail::deferred_config_element_base in order to satisfy
+ * this concept. This concept is merely used for internal purposes and shall not be exposed in public user interfaces.
+ */
+//!\cond
+template <typename config_t>
+concept bool deferred_config_element_concept =
+    std::is_base_of_v<deferred_config_element_base<std::remove_reference_t<config_t>>, std::remove_reference_t<config_t>> &&
+    config_element_concept<config_t> &&
+    invocable_concept<std::remove_reference_t<config_t>,
+                      decltype(_callable_dummy),
+                      detail::configuration<std::remove_reference_t<config_t>>>;
```

Comment:
when would this be false?

Reply:
If any of the conditions is false?
But of course since I require the base of I can drop the imvocable. Ah that’s what you meant. Well, I wasn’t sure about having a concept requiring a base class of the type🙄🤖. Then again it is only to use forwarding references in internal function declarations

Reply:
yeah, I was referring to the invocable, it's confusing me :laughing: